### PR TITLE
feat: support classification

### PR DESCRIPTION
**Pull request by n0w0f**
There is lot of duplications, so as to avoid any breakage at this moment.
might have to refactor later.

we could also choose to not merge it. but would be good to review the code.

## Summary by Sourcery

Add support for classification tasks by introducing new classes and methods, refactor benchmarking and task execution logic for better modularity, and enhance model fine-tuning and inference processes with abstract base classes. Include new configuration files and scripts for data preparation and model setup.

New Features:
- Introduce support for classification tasks by adding new classes and methods for handling classification-specific operations.
- Add new configuration files for various model representations and datasets, enabling flexible model training and evaluation setups.

Enhancements:
- Refactor the benchmarking and task running logic to use a more modular and abstract base class structure, improving code organization and reusability.
- Enhance the task runner to dynamically map and execute tasks based on a predefined task map, allowing for more flexible task execution.
- Improve the model fine-tuning and inference processes by introducing abstract base classes and methods, facilitating easier extension and customization.

Documentation:
- Add new configuration files for different model setups, providing examples and templates for various training scenarios.

Chores:
- Add multiple new configuration files and scripts for data preparation and processing, supporting the new classification feature and other model setups.

### CODE REVIEW

**Comment by n0w0f** [comment]
@sourcery-ai review 
Can you check if the roc_auc and metric computation is done correctly in `score.py`

**Comment by kjappelbaum** [comment]
As you and Sourcery said, there is a lot of duplication, and I'm unsure if we should merge it in the current form.

File: revision-scripts/prep_rep.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        return text_reps  # Return the entire dictionary
+    except TimeoutError:
+        print("Timeout error processing a row")
+        return None
+    except Exception as e:
+        print(f"Error processing a row: {e}")
+        return None
+
+    
+def process_entry_test_matbench(entry: List, timeout: int) -> dict:
```
Comment (kjappelbaum) [inline revision-scripts/prep_rep.py:44]: we should also try to revise functions like this one later - e.g. `timeout` seems to not be used

File: revision-scripts/prep_rep.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    data = read_json(json_file)
+    batch_size = num_workers * 4
+
+    if last_processed_entry > 0:
+        data = data[last_processed_entry:]
+
+    batch_iterator = (data[i:i + batch_size] for i in range(0, len(data), batch_size))
+
+    for i, batch_data in enumerate(batch_iterator, start=1):
```
Comment (kjappelbaum) [inline revision-scripts/prep_rep.py:96]: why do you actually start at 1 and not 0?

File: src/mattext/main.py [hunk 3]
```diff
@@ -22,6 +22,9 @@ def __init__(self):
     def run_task(self, run: list, task_cfg: DictConfig, local_rank=None) -> None:
         if "benchmark" in run:
             self.run_benchmarking(task_cfg)
+ 
+        if "classification" in run:
+            self.run_classification(task_cfg)
```
Comment (kjappelbaum) [inline src/mattext/main.py:27]: also this is technical debt we should remove - in the end, it is just a special case of benchmarking

File: src/mattext/models/classification.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        precision, recall, f1, _ = precision_recall_fscore_support(labels, preds_argmax, average='weighted')
+        acc = accuracy_score(labels, preds_argmax)
+
+        # Compute ROC AUC
+        n_classes = preds.shape[1]
+        if n_classes == 2:
+            # Binary classification
+            roc_auc = roc_auc_score(labels, preds[:, 1])
+        else:
+            # Multi-class classification
```
Comment (kjappelbaum) [inline src/mattext/models/classification.py:117]: what is the multiclass classification you deal with? `is_metal` seems binary? Or is this just for "future proofing"?

Comment (kjappelbaum) [reply src/mattext/models/classification.py:117]: and you need to slice out one class because your model predicts the probability for True and False?

Comment (kjappelbaum) [reply src/mattext/models/classification.py:117]: I see, you have `num_labels` hardcoded  - in principle, one should be enough for binary

**Comment by kjappelbaum** [comment]
> @sourcery-ai review Can you check if the roc_auc and metric computation is done correctly in `score.py` ?

Is there anything particular you have concerns about, @n0w0f ?

**Comment by n0w0f** [comment]
> > @sourcery-ai review Can you check if the roc_auc and metric computation is done correctly in `score.py` ?
> 
> Is there anything particular you have concerns about, @n0w0f ?

I wanted to jsut confirm if the metrics are correctly computed.

**Comment by n0w0f** [comment]
@kjappelbaum I agree. we need not merge this. I did not want to break the codebase hence the duplications I can clean up after the sprint

**Comment by kjappelbaum** [comment]
i'd also consider moving to something like unsloth for fine-tuning as it is much faster

**Comment by kjappelbaum** [comment]
Another point we had mentioned on Zulip was to add 

- batching 
- progress bars

File: src/mattext/models/score.py [hunk 5]
```diff
@@ -131,3 +142,89 @@ def _json_serializable(obj):
...
+    def _json_serializable(obj):
+        if isinstance(obj, (np.ndarray, pd.Series)):
+            return obj.tolist()
+        elif isinstance(obj, np.bool_):
+            return bool(obj)
+        elif isinstance(obj, np.integer):
+            return int(obj)
+        elif isinstance(obj, np.floating):
+            return float(obj)
+        raise TypeError(f"Type {type(obj)} not serializable")
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/score.py]: done

File: src/mattext/models/benchmark.py [hunk 6]
```diff
@@ -128,3 +128,114 @@ def run_benchmarking(self, local_rank=None) -> None:
...
+                predictions, prediction_ids = inference.predict()
+                # task.record_fold(fold=i, prediction_ids=prediction_ids, predictions=predictions.values)
+                print("--------------------")
+                print(len(prediction_ids), len(predictions))
+                print("---------------------")
+                task.record_fold(
+                    fold=i, prediction_ids=prediction_ids, predictions=predictions.values
+                )
+
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py]: done

File: src/mattext/main.py [hunk 7]
```diff
@@ -22,6 +22,9 @@ def __init__(self):
     def run_task(self, run: list, task_cfg: DictConfig, local_rank=None) -> None:
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/main.py:34]: done

File: src/mattext/models/predict.py [hunk 8]
```diff
@@ -81,3 +88,131 @@ def predict(self):
         self.prediction_ids = prediction_ids
 
         return pd.Series(predictions.predictions.flatten()), prediction_ids
+
+
+class InferenceClassification(TokenizerMixin):
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/predict.py]: abstracted out common part to a base class. can you check

File: src/mattext/models/benchmark.py [hunk 9]
```diff
@@ -128,3 +128,114 @@ def run_benchmarking(self, local_rank=None) -> None:
         # Get final results after recording all folds
         # final_results = task.get_final_results()
         # print(final_results)
+
+
+class MatbenchmarkClassification:
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py]: added polymorphism

**Comment by n0w0f** [comment]
@sourcery-ai review

File: src/mattext/models/score.py [hunk 10]
```diff
@@ -1,58 +1,51 @@
...
 
 
-def mattext_score(prediction_ids, predictions, task_name):
-    true = load_true_scores(task_name, prediction_ids)
-    return mean_squared_error(true, predictions)
-
-
 @dataclass
-class MatTextTask:
+class BaseMatTextTask:
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/score.py]: combined to single a single flexible class. can you check the latest commits

File: src/mattext/models/benchmark.py [hunk 11]
```diff
@@ -39,92 +28,138 @@
...
         else:
             task = MatTextTask(task_name=self.task)
+        return task
+
+    def _run_experiment(self, task, i, exp_name, test_name, local_rank):
+        fold_name = fold_key_namer(i)
+        print(
+            f"Running training on {self.train_data}, and testing on {self.test_data} for fold {i}"
+        )
+        print("-------------------------")
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py:50]: there is only one instance of this print

File: revision-scripts/mp_classification.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                             readahead=False,
+                             meminit=False,
+                             max_readers=max_readers)
+        self.txn = self.env.begin()
+
+    def __len__(self):
+        return self.txn.stat()['entries']
+
+    def get(self, index):
+        id = f"{index}".encode("ascii")
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply revision-scripts/mp_classification.py:23]: replaced `id` with `id_`

File: src/mattext/models/benchmark.py [hunk 13]
```diff
@@ -39,92 +28,138 @@
...
+
+        try:
+            predict = self._get_inference(exp_cfg, fold_name)
+            predictions, prediction_ids = predict.predict()
+            self._record_predictions(task, i, predictions, prediction_ids)
+        except Exception as e:
+            print(
+                f"Error occurred during inference for finetuned checkpoint '{exp_name}': {str(e)}"
+            )
+            if isinstance(e, (ValueError, TypeError)):
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py:81]: could be a different PR. logger is not used yet

File: src/mattext/models/benchmark.py [hunk 14]
```diff
@@ -1,34 +1,23 @@
...
 
-class Matbenchmark:
-    """
-    Class to perform predictions on Matbench datasets.
-
-    Args:
-    - task_cfg (DictConfig): Configuration dictionary containing task parameters.
-    """
-
+class BaseBenchmark(ABC):
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py:18]: common methods is in base class already, maybe check latest commits

File: src/mattext/models/predict.py [hunk 15]
```diff
@@ -81,3 +88,131 @@ def predict(self):
...
+            special_num_token=cfg.model.special_num_token,
+        )
+        self.fold = fold
+        self.representation = cfg.model.representation
+        self.data_repository = cfg.model.data_repository
+        self.dataset_name = cfg.model.finetune.dataset_name
+        self.cfg = cfg.model.inference
+        self.context_length: int = self.cfg.context_length
+        #self.num_labels = cfg.model.num_labels
+        self.num_labels = 2
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/predict.py]: will be addressed later, feel free to create a new issue

File: revision-scripts/mp_classification.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        datapoint = pickle.loads(self.txn.get(id))
+        return datapoint
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply revision-scripts/mp_classification.py]: done

File: src/mattext/models/benchmark.py [hunk 17]
```diff
@@ -57,7 +58,7 @@
             Exception: If an error occurs during inference for a finetuned checkpoint.
 
         """
-        if self.task_type == "matbench":
+        if self.task_type == "matbench" or self.task_type == "classification":
```
[the remark this answers is not in the document]
Comment (n0w0f) [reply src/mattext/models/benchmark.py]: not applicable for the new changes

File: src/mattext/models/classification.py (hunk 4 above)
Comment (n0w0f) [reply src/mattext/models/classification.py]: for now hardcoding num_labels to two.
could be read from config later

**Comment by n0w0f** [comment]
Refactored to incorporate the comments. can you @sourcery-ai review

**Pull request merged**